cmd/soroban-rpc/internal/methods/get_events.go

```diff
@@ -0,0 +1,475 @@
+	if len(g.Filters) > 5 {
+		return errors.New("maximum 5 filters per request")
+	}
+	for i, filter := range g.Filters {
+		if err := filter.Valid(); err != nil {
+			return errors.Wrapf(err, "filter %d invalid", i+1)
+		}
+	}
+
+	return nil
+}
+
+func (g *GetEventsRequest) Matches(event xdr.ContractEvent) bool {
+	if len(g.Filters) == 0 {
+		return true
+	}
+	for _, filter := range g.Filters {
+		if filter.Matches(event) {
+			return true
+		}
+	}
+	return false
+}
+
+type EventFilter struct {
+	EventType   string        `json:"type,omitempty"`
+	ContractIDs []string      `json:"contractIds,omitempty"`
+	Topics      []TopicFilter `json:"topics,omitempty"`
+}
+
+func (e *EventFilter) Valid() error {
+	switch e.EventType {
+	case "", "system", "contract":
+		// ok
+	default:
+		return errors.New("if set, type must be either 'system' or 'contract'")
+	}
+	if len(e.ContractIDs) > 5 {
+		return errors.New("maximum 5 contract IDs per filter")
+	}
+	if len(e.Topics) > 5 {
+		return errors.New("maximum 5 topics per filter")
+	}
+	for i, id := range e.ContractIDs {
+		out, err := hex.DecodeString(id)
+		if err != nil || len(out) != 32 {
+			return fmt.Errorf("contract ID %d invalid", i+1)
+		}
+	}
+	for i, topic := range e.Topics {
+		if err := topic.Valid(); err != nil {
+			return errors.Wrapf(err, "topic %d invalid", i+1)
+		}
+	}
+	return nil
+}
+
+// TODO: Implement this more efficiently (ideally do it in the real data backend)
+func (e *EventFilter) Matches(event xdr.ContractEvent) bool {
+	return e.matchesEventType(event) && e.matchesContractIDs(event) && e.matchesTopics(event)
+}
+
+func (e *EventFilter) matchesEventType(event xdr.ContractEvent) bool {
+	if e.EventType == "contract" && event.Type != xdr.ContractEventTypeContract {
+		return false
+	}
+	if e.EventType == "system" && event.Type != xdr.ContractEventTypeSystem {
+		return false
+	}
+	return true
+}
+
+func (e *EventFilter) matchesContractIDs(event xdr.ContractEvent) bool {
+	if len(e.ContractIDs) == 0 {
+		return true
+	}
+	if event.ContractId == nil {
+		return false
+	}
+	needle := hex.EncodeToString((*event.ContractId)[:])
+	for _, id := range e.ContractIDs {
+		if id == needle {
+			return true
+		}
+	}
+	return false
+}
+
+func (e *EventFilter) matchesTopics(event xdr.ContractEvent) bool {
+	if len(e.Topics) == 0 {
+		return true
+	}
+	v0 := event.Body.MustV0()
+	for _, topicFilter := range e.Topics {
+		if topicFilter.Matches(v0.Topics) {
+			return true
+		}
+	}
+	return false
+}
+
+type TopicFilter []SegmentFilter
+
+func (t *TopicFilter) Valid() error {
+	if len(*t) < 1 {
+		return errors.New("topic must have at least one segment")
+	}
+	if len(*t) > 4 {
+		return errors.New("topic cannot have more than 4 segments")
+	}
+	return nil
+}
+
+func (t TopicFilter) Matches(event []xdr.ScVal) bool {
+	for _, segmentFilter := range t {
+		if segmentFilter.wildcard != nil {
+			switch *segmentFilter.wildcard {
+			case "*":
+				// one-segment wildcard
+				if len(event) == 0 {
```

Comment:
just for consideration, i think could move this guard check on lines 180 - 183 up to top of loop, after 175, and then delete the duplicate check from lines 191-193.